azure-proto-core/ApiVersionsBase.cs

```diff
@@ -0,0 +1,72 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace azure_proto_core
+{
+    public class ApiVersionsBase : IEquatable<string>, IComparable<string>
+    {
+        private string _value;
+
+        protected ApiVersionsBase(string value)
+        {
+            _value = value;
+        }
+
+        public static implicit operator string(ApiVersionsBase version)
+        {
+            return version._value;
+        }
+
+        public static bool operator ==(ApiVersionsBase first, string second)
+        {
+            if (ReferenceEquals(null, first))
+                return ReferenceEquals(null, second);
+            if (ReferenceEquals(null, second))
+                return false;
+            return first.Equals(second);
+        }
+
+        public static bool operator !=(ApiVersionsBase first, string second)
+        {
+            if (ReferenceEquals(null, first))
+                return !ReferenceEquals(null, second);
+            if (ReferenceEquals(null, second))
+                return true;
+            return !first.Equals(second);
+        }
+
+        public override string ToString()
+        {
+            return _value;
+        }
+
+        public override bool Equals(object obj)
+        {
+            if (obj is ApiVersionsBase)
+                return Equals(obj as ApiVersionsBase);
```

Comment:
I don't see an overload of Equals overload that take an ApiVersionBase. What method does this end up calling? I'm not sure if it recursively calls this method or if it calls the string overload?

Reply:
It calls the string version using the implicit.  This avoids ambiguous method error if a consumer uses
ApiVersionBase version;
versioin.Equals(null);

Reply:
Perhaps we should make this explicit and do `Equals(obj as string)` here as well.

Reply:
I do have that

            if (obj is ApiVersionsBase)
                return Equals(obj as ApiVersionsBase);
            if (obj is string)
                return Equals(obj as string);